### ISSUE TITLE
[1.1] Port 2-3 scrapers to Memorious / Scrapy

### ISSUE THREAD

**Issue by robredpath** [open-contracting/kingfisher#262]
Choose some that look quite different / use different patterns so that we can try out the new framework in a few different ways. 

Work on this (should!) be able to start on Friday. A co-ordinating call/email/ticket/issue/WhatsApp is probably a good idea!

**Comment by odscjames** [comment]
A lot of scrapers follow the same patterns. So when picking which scrapers to work on, try and do one of each of the items in the list first. 

* Hardcoded list of URLs: canada_buyandsell, honduras_cost, taiwan, 
* Paging by number; Make one request to get total number of pages, can generate a full list of URLS straight away: canada_montreal, colombia, mexico_administracion_publica_federal, uk_contracts_finder, 
* Make request to get a list of data, maybe loop over dates, maybe that list is also a list of lists so more requests needed, eventually you have a list of URLs: afghanistan_records, afghanistan_releases, chile_compra, mexico_cdmx, mexico_grupo_aeroporto, moldova
* Like the last one, but sometimes valid OCDS data links to other OCDS data: mexico_jalisco, uganda
* Paging by next: Every time you make a request, you discover an extra URL to go on the list: armenia
* Special Cases: mexico_inai - build URLs from a redirect
* Special Cases: Paraguay_dncp, paraguay_hacienda - You get to make an API token, and caching that token would be good
* Special Cases: scotland - Looping 14 days at a time 
* Special Cases: Ukraine - Scraping HTML for links.
* Special Cases: Uruguay - A RSS reader
* Special Cases: Zambia - Uncompress a ZIP file.
* Special Cases: digiwhist_* - Uncompress a ZIP file …. Of JSON Lines!

 (Note some scrapers aren't above for various reasons: australia_nsw,  indonesia_bandung)

**Comment by odscjames** [comment]
The special cases just means there is only one of that type. Out of the special cases, digiwhist_* is probably the most important. We currently have problems with good error reporting around JSON Lines so it would be good to look at that soon.

**Comment by yolile** [comment]
So just to put here what we agree in the call: 
- Amy is going to do the Memorious scrappers and I'm going to do the Scrapy ones
- we are going to do the scrapers for: canada_buyandsell, afghnistan and paraguay_dncp

**Comment by jpmckinney** [comment]
I thought we'd have a discussion about #261 before doing trial implementations. There are many considerations regarding Memorious vs Scrapy vs Status quo that aren't technical considerations. Trial implementations will only answer technical questions. I can write my thoughts on those if the decision has been to disregard such considerations and to focus on the technical level only.

I see no new notes in the [sprint documentation](https://docs.google.com/document/d/1zeBhhSV9eLPoRu8cIy-0y4HNT02fCXNr3b6LdtyijVM/edit), so I'm not sure why the decision is to do trial implementations.

Please invite me to calls, so that we don't have to stop-and-start work. I might not join every call, but it should help avoid this sort of back-and-forth.

**Comment by robredpath** [comment]
@jpmckinney this issue and #261 are two sides of the same coin, and are about far more than just the technical questions. By trying to implement some scrapers, we get an idea of what the workflow and features of each framework looks like, and what they're like for people to work with, in the context of working with OCDS data and the sorts of APIs/websites that we deal with in OCDS-world. It's common for people to find it easier to learn how to use tools and how to assess them by having a real-world problem to solve, and writing scrapers with an eye to thinking about how they'd be used in production feels like a good real-world problem to focus our attention. 

Yesterday's call was to turn the sprint coordination doc into actionable tickets, which allow more focussed collaboration around specific issues. If you've got thoughts on the Memorious vs Scrapy vs status quo, then please do share them here or in #261 so that we can proceed. I'm keen to make sure that we know in the next few days whether a wholesale re-architecting is worthwhile and what it looks like, so that we can spend January either implementing it, or building ways to achieve the agreed deliverables with the existing architecture.

**Comment by jpmckinney** [comment]
I added comments to #261. I'm in favor of learning by doing – however, I'm not sure the experimental design here is appropriate. I'm concerned that we won't learn anything except that Amy is good at writing scrapers with Memorious, and that Yohanna is good with Scrapy.

I think it'd be more appropriate and efficient to (1) assess, as well as we could, which of the two to use (#261); (2) try using that one (this issue); (3) if it seems we made an error in (1) based on experience in (2), try using the other one. If we run into no problems in (2), then we save time doing any work with the other one.

What's important here in terms of experimental design is that in (2), we'll be working as a team to figure out any barriers we hit: we'll be doing our best to make it work; we'll be checking that any issue we encounter is truly a problem with the tool, not with our understanding of it; and we'll know that the problem is with the tool, and not with one developer having a bad day or whatnot.

An even more considered experimental design would set out specific questions we hope to answer through the test, based on our preliminary evaluation of of the two in (1). We've already started on this by having the target be three different scraper patterns. If we're uncertain that both can handle those patterns, then it's a good question (though, do we really think that this is going to be an issue for either Scrapy or Memorious?). We might also (or alternately) have the targets be:
* Accept a parameter that changes the scraper's storage method (viz. https://github.com/open-contracting/kingfisher/issues/261#issuecomment-447129791)
* Report HTTP/API errors for a source that is known to produce such errors
* Report the scraper's activity, including an indication of progress

Anyway, in short, why write scrapers for both, instead of one at a time, starting with the one we feel is most likely to be the right choice?

----

For critical moments like turning the coordination document into actionable tickets, I think it's important for me to listen in on the calls, so that where appropriate we can have these discussions at that time, instead of in GitHub issues.

**Comment by yolile** [comment]
Just one example of the canada source using scrapy is available here: https://github.com/yolile/kingfisher_scrapy

@odscjames you can see there how you can pass parameters as filename, data_type, etc and also how to use the command line arguments, for example 'sample'
cc @rhiaro @BibianaC

**Comment by BibianaC** [comment]
Thanks @yolile. Do you think that 'Feed exports' could help to generate a file?
https://docs.scrapy.org/en/latest/topics/feed-exports.html#topics-feed-exports

**Comment by robredpath** [comment]
https://github.com/open-contracting/kingfisher-scrape is where these live, on the master-scrapy branch

**Comment by jpmckinney** [comment]
More than 2-3 scrapers written, so closing https://github.com/open-contracting/kingfisher-scrape/tree/master/kingfisher_scrapy/spiders
